### ISSUE TITLE
Jupyterlab on Jupyterhub cannot interact with many settings buttons (Returns 400 on PUT)

### ISSUE THREAD

**Issue by OneRaynyDay** [jupyterlab/jupyterlab#7162]
## Description

Hi, I am currently trying to use functionalities from the settings tab, such as the Extension Manager. I can't seem to open it as I get a 400 from a PUT request to https://<my_url>/user/ray_zhang/lab/api/settings/@jupyterlab/extensionmanager-extension:plugin?1567806980838 

Here is the error: 
![image](https://user-images.githubusercontent.com/7191678/64463221-11a38600-d0b8-11e9-9694-7ee8fc4fe332.png)

Tracing down the stacktrace, I see the following:

In `serverconnection.js`:
```
    /**
     * Make an request to the notebook server.
     *
     * @param url - The url for the request.
     *
     * @param init - The initialization options for the request.
     *
     * @param settings - The server settings to apply to the request.
     *
     * @returns a Promise that resolves with the response.
     *
     * @throws If the url of the request is not a notebook server url.
     *
     * #### Notes
     * The `url` must start with `settings.baseUrl`.  The `init` settings are
     * merged with `settings.init`, with `init` taking precedence.
     * The headers in the two objects are not merged.
     * If there is no body data, we set the content type to `application/json`
     * because it is required by the Notebook server.
     */
    function makeRequest(url, init, settings) {
        return Private.handleRequest(url, init, settings); // <-- errors here, line 75
    }
...
    class ResponseError extends Error {
        /**
         * Create a new response error.
         */
        constructor(response, message) {
            message =
                message ||
                    `Invalid response: ${response.status} ${response.statusText}`;
            super(message); // error here <-- line 89
            this.response = response;
        }
    }
```

In `settingregistry.js`:
```
    /**
     * Save a plugin in the registry.
     */
    async _save(plugin) {
        const plugins = this.plugins;
        if (!(plugin in plugins)) {
            throw new Error(`${plugin} does not exist in setting registry.`);
        }
        try {
            await this._validate(plugins[plugin]);
        }
        catch (errors) {
            console.warn(`${plugin} validation errors:`, errors);
            throw new Error(`${plugin} failed to validate; check console.`);
        }
        await this.connector.save(plugin, plugins[plugin].raw); // <-- fails here, line 392
        // Fetch and reload the data to guarantee server and client are in sync.
        const fetched = await this.connector.fetch(plugin);
        await this._load(await this._transform('fetch', fetched));
        this._pluginChanged.emit(plugin);
    }
```

In `index.js`:
```
/**
     * Save a plugin's settings.
     *
     * @param id - The plugin's ID.
     *
     * @param raw - The user setting values as a raw string of JSON with comments.
     *
     * @returns A promise that resolves if successful.
     */
    async save(id, raw) {
        const { serverSettings } = this;
        const { baseUrl, appUrl } = serverSettings;
        const { makeRequest, ResponseError } = serverconnection_1.ServerConnection;
        const base = baseUrl + appUrl;
        const url = Private.url(base, id);
        const init = { body: raw, method: 'PUT' };
        const response = await makeRequest(url, init, serverSettings); // <-- fails here, line 84
        if (response.status !== 204) {
            throw new ResponseError(response);
        }
    }
```

The effect of this 400 is that the dropdown selection is unresponsive, and no manager shows up. This also occurs for changing the keybindings to vim, changing theme to dark, etc, all with similar codepaths I believe.

## Reproduce

1. Go to 'Settings' tab
2. Click on 'Enable Extension Manager (experimental)'
3. See no response. Alternatively, open developer's console and observe 400's.

## Context

Running inside Docker, running ubuntu 18.04. The host instance is AWS EC2, ubuntu 14.04.

- Operating System and version:
- Browser and version: Google Chrome, Version 76.0.3809.132 (Official Build) (64-bit)
- JupyterLab version:

```
jupyterlab==1.0.5
jupyterlab-server==1.0.6
```

<details><summary>Troubleshoot Output</summary>
<pre>
$PATH:
	/opt/conda/bin
	/usr/local/nvm/versions/node/v12.8.1/bin
	/usr/local/sbin
	/usr/local/bin
	/usr/sbin
	/usr/bin
	/sbin
	/bin
	/srv/lib/spark/spark-2.4.0-bin-hadoop2.7/bin

sys.path:
	/opt/conda/envs/jupyterhub/bin
	/opt/conda/envs/jupyterhub/lib/python36.zip
	/opt/conda/envs/jupyterhub/lib/python3.6
	/opt/conda/envs/jupyterhub/lib/python3.6/lib-dynload
	/opt/conda/envs/jupyterhub/lib/python3.6/site-packages

sys.executable:
	/opt/conda/envs/jupyterhub/bin/python

sys.version:
	3.6.9 |Anaconda, Inc.| (default, Jul 30 2019, 19:07:31)
	[GCC 7.3.0]

platform.platform():
	Linux-4.4.0-159-generic-x86_64-with-debian-buster-sid

pip list:
	Package            Version
	------------------ ---------
	alembic            0.9.3
	asn1crypto         0.24.0
	attrs              19.1.0
	backcall           0.1.0
	bleach             3.1.0
	boto3              1.4.4
	botocore           1.5.95
	certifi            2019.6.16
	cffi               1.12.3
	chardet            3.0.4
	cryptography       2.7
	datadog            0.28.0
	decorator          4.4.0
	deepmerge          0.0.4
	defusedxml         0.6.0
	docker             2.4.2
	docker-pycreds     0.4.0
	docutils           0.15.2
	entrypoints        0.3
	escapism           1.0.0
	idna               2.5
	ipykernel          5.1.2
	ipython            7.8.0
	ipython-genutils   0.2.0
	ipywidgets         7.5.0
	jedi               0.15.1
	Jinja2             2.10.1
	jmespath           0.9.4
	json5              0.8.5
	jsonschema         3.0.2
	jupyter-client     5.3.1
	jupyter-console    5.2.0
	jupyter-core       4.5.0
	jupyterhub         0.7.2
	jupyterlab         1.0.5
	jupyterlab-server  1.0.6
	Mako               1.1.0
	MarkupSafe         1.1.1
	mistune            0.8.4
	mysqlclient        1.3.12
	nbconvert          5.6.0
	nbformat           4.4.0
	nose               1.3.7
	notebook           6.0.1
	pamela             0.3.0
	pandocfilters      1.4.2
	parso              0.5.1
	pexpect            4.7.0
	pickleshare        0.7.5
	pip                19.2.2
	prometheus-client  0.7.1
	prompt-toolkit     2.0.9
	ptyprocess         0.6.0
	pycparser          2.19
	Pygments           2.4.2
	pyhocon            0.3.38
	pyOpenSSL          17.4.0
	pyparsing          2.4.2
	pyrsistent         0.15.4
	python-dateutil    2.8.0
	python-editor      1.0.4
	PyYAML             3.12
	pyzmq              18.1.0
	qtconsole          4.5.1
	raven              6.10.0
	requests           2.18.1
	s3transfer         0.1.13
	Send2Trash         1.5.0
	setuptools         41.2.0
	six                1.12.0
	SQLAlchemy         1.1.11
	straitlets         0.3.1
	terminado          0.8.2
	testpath           0.4.2
	tornado            4.5.1
	traitlets          4.3.2
	urllib3            1.21.1
	wcwidth            0.1.7
	webencodings       0.5.1
	websocket-client   0.56.0
	wheel              0.33.4
	widgetsnbextension 3.5.1

conda list:
	# packages in environment at /opt/conda:
	#
	asn1crypto                0.22.0           py27h94ebe91_1
	ca-certificates           2017.08.26           h1d4fec5_0
	certifi                   2017.7.27.1      py27h9ceb091_0
	cffi                      1.10.0           py27hf1aaaf4_1
	chardet                   3.0.4            py27hfa10054_1
	conda                     4.3.27           py27hff99c7a_0
	conda-env                 2.6.0                h36134e3_1
	cryptography              2.0.3            py27hea39389_1
	enum34                    1.1.6            py27h99a27e9_1
	idna                      2.6              py27h5722d68_1
	ipaddress                 1.0.18           py27h337fd85_0
	libedit                   3.1                  heed3624_0
	libffi                    3.2.1                h4deb6c0_3
	libgcc-ng                 7.2.0                hcbc56d2_1
	libstdcxx-ng              7.2.0                h24385c6_1
	ncurses                   6.0                  h06874d7_1
	openssl                   1.0.2l               h9d1a558_3
	pip                       9.0.1            py27h4c18a59_2
	pycosat                   0.6.2            py27h1cf261c_1
	pycparser                 2.18             py27hefa08c5_1
	pyopenssl                 17.2.0           py27h189ff3b_0
	pysocks                   1.6.7            py27he2db6d2_1
	python                    2.7.13              hfff3488_13
	readline                  7.0                  hac23ff0_3
	requests                  2.18.4           py27hc5b0589_1
	ruamel_yaml               0.11.14          py27h672d447_2
	setuptools                36.5.0           py27h68b189e_0
	six                       1.10.0           py27hdcd7534_1
	sqlite                    3.20.1               h6d8b0f3_1
	tk                        8.6.7                h5979e9b_1
	urllib3                   1.22             py27ha55213b_0
	wheel                     0.29.0           py27h411dd7b_1
	yaml                      0.1.7                h96e3832_1
	zlib                      1.2.11               hfbfcf68_1
</pre>
</details>

<details><summary>Command Line Output</summary>
<pre>
[D 22:13:07.120 LabApp] Searching ['/opt/conda/envs/jupyterhub/bin', '/efs/home/ray_zhang/.jupyter', '/opt/conda/envs/jupyterhub/etc/jupyter', '/usr/local/etc/jupyter', '/etc/jupyter'] for config files
[D 22:13:07.120 LabApp] Looking for jupyter_config in /etc/jupyter
[D 22:13:07.121 LabApp] Looking for jupyter_config in /usr/local/etc/jupyter
[D 22:13:07.121 LabApp] Looking for jupyter_config in /opt/conda/envs/jupyterhub/etc/jupyter
[D 22:13:07.121 LabApp] Looking for jupyter_config in /efs/home/ray_zhang/.jupyter
[D 22:13:07.125 LabApp] Looking for jupyter_config in /opt/conda/envs/jupyterhub/bin
[D 22:13:07.127 LabApp] Looking for jupyter_notebook_config in /etc/jupyter
[D 22:13:07.127 LabApp] Loaded config file: /etc/jupyter/jupyter_notebook_config.py
[D 22:13:07.128 LabApp] Looking for jupyter_notebook_config in /usr/local/etc/jupyter
[D 22:13:07.128 LabApp] Looking for jupyter_notebook_config in /opt/conda/envs/jupyterhub/etc/jupyter
[D 22:13:07.128 LabApp] Looking for jupyter_notebook_config in /efs/home/ray_zhang/.jupyter
[D 22:13:07.141 LabApp] Loaded config file: /efs/home/ray_zhang/.jupyter/jupyter_notebook_config.json
[D 22:13:07.141 LabApp] Looking for jupyter_notebook_config in /opt/conda/envs/jupyterhub/bin
[W 22:13:07.143 LabApp] server_extensions is deprecated, use nbserver_extensions
[D 22:13:07.300 LabApp] Paths used for configuration of notebook:
    	/efs/home/ray_zhang/.jupyter/nbconfig/notebook.json
[D 22:13:07.316 LabApp] Paths used for configuration of notebook:

[D 22:13:07.349 LabApp] Paths used for configuration of notebook:
    	/efs/home/ray_zhang/.jupyter/nbconfig/notebook.json
[D 22:13:07.357 LabApp] Paths used for configuration of notebook:

[D 22:13:07.389 LabApp] Paths used for configuration of notebook:
    	/efs/home/ray_zhang/.jupyter/nbconfig/notebook.json
[D 22:13:07.397 LabApp] Paths used for configuration of notebook:

[D 22:13:07.514 LabApp] Paths used for configuration of jupyter_notebook_config:
    	/etc/jupyter/jupyter_notebook_config.json
[D 22:13:07.515 LabApp] Paths used for configuration of jupyter_notebook_config:
    	/usr/local/etc/jupyter/jupyter_notebook_config.json
[D 22:13:07.515 LabApp] Paths used for configuration of jupyter_notebook_config:
    	/opt/conda/envs/jupyterhub/etc/jupyter/jupyter_notebook_config.d/jupyterlab.json
    	/opt/conda/envs/jupyterhub/etc/jupyter/jupyter_notebook_config.json
[D 22:13:07.518 LabApp] Paths used for configuration of jupyter_notebook_config:
    	/efs/home/ray_zhang/.jupyter/jupyter_notebook_config.json
[I 22:13:07.531 LabApp] JupyterLab extension loaded from /opt/conda/envs/jupyterhub/lib/python3.6/site-packages/jupyterlab
[I 22:13:07.531 LabApp] JupyterLab application directory is /opt/conda/envs/jupyterhub/share/jupyter/lab
[W 22:13:07.533 LabApp] JupyterLab server extension not enabled, manually loading...
[I 22:13:07.537 LabApp] JupyterLab extension loaded from /opt/conda/envs/jupyterhub/lib/python3.6/site-packages/jupyterlab
[I 22:13:07.537 LabApp] JupyterLab application directory is /opt/conda/envs/jupyterhub/share/jupyter/lab
[C 22:13:07.538 LabApp] Running as root is not recommended. Use --allow-root to bypass.
[D 22:13:07.538 LabApp] Exiting application: jupyter-notebook
</pre>
</details>

**Comment by jasongrout** [comment]
#8221 documents the issue, but let's leave this open until we actually take care of the underlying issue. More debugging at #8221 shows where the problem is.

**Comment by jasongrout** [comment]
Basically, the problem here ended up being that JupyterLab claims to be sending JSON when saving settings, but really is sending JSON5, so devices on the network may throw errors when trying to validate the payload as JSON. See #8221 for more details.

**Comment by jasongrout** [comment]
Fixed by #8225
